samples/guide/src/main/java/okhttp3/recipes/kt/WiresharkExample.kt

```diff
@@ -0,0 +1,246 @@
+      // Consuming server Finished handshake message
+      // Produced ClientHello handshake message
+      //
+      // Raw write
+      // Raw read
+      // Plaintext before ENCRYPTION
+      // Plaintext after DECRYPTION
+      val message = record.message
+
+      val parameters = record.parameters
+
+      if (parameters != null && !message.startsWith("Raw") && !message.startsWith("Plaintext")) {
+//        p(record.message)
+//        p(record.parameters[0])
+        val param = parameters[0] as String
+
+        if (message == "Produced ClientHello handshake message") {
+          random = readClientRandom(param)
+        }
+      }
+    }
+
+    override fun flush() {}
+
+    override fun close() {}
+  }
+
+  private fun readClientRandom(param: String): String? {
+    val matchResult = randomRegex.find(param)
+
+    return if (matchResult != null) {
+      matchResult.groupValues[1].replace(" ", "")
+    } else {
+      null
+    }
+  }
+
+  override fun secureConnectStart(call: Call) {
+    // Register to capture "Produced ClientHello handshake message".
+
+    currentThread = Thread.currentThread()
+    logger.addHandler(loggerHandler)
+  }
+
+  override fun secureConnectEnd(
+    call: Call,
+    handshake: Handshake?
+  ) {
+    logger.removeHandler(loggerHandler)
+  }
+
+  override fun callEnd(call: Call) {
+    // Cleanup log handler if failed.
+    logger.removeHandler(loggerHandler)
+  }
+
+  override fun connectionAcquired(
+    call: Call,
+    connection: Connection
+  ) {
+    if (random != null) {
+      val sslSocket = connection.socket() as SSLSocket
+      val session = sslSocket.session
+
+      val masterSecretHex = session.masterSecret?.encoded?.toByteString()
+          ?.hex()
+//          p(masterSecretHex)
+//          p(session.id.toByteString().hex())
+
+      if (masterSecretHex != null) {
+        val keyLog = "CLIENT_RANDOM $random $masterSecretHex"
+
+        println(keyLog)
+        File("/tmp/key.log").appendText("$keyLog\n")
+      }
+    }
+
+    random = null
+  }
+
+  class Factory : EventListener.Factory {
+    override fun create(call: Call): EventListener {
+      return WireSharkKeyLoggerListener()
+    }
+  }
+
+  companion object {
+    private lateinit var logger: Logger
+
+    private val SSLSession.masterSecret: SecretKey?
+      get() = javaClass.getDeclaredField("masterSecret")
+          .apply {
+            isAccessible = true
+          }
+          .get(this) as? SecretKey
+
+    val randomRegex = "\"random\"\\s+:\\s+\"([^\"]+)\"".toRegex()
+
+    fun register() {
+      // Enable JUL logging for SSL events.
+      System.setProperty("javax.net.debug", "")
+      logger = Logger.getLogger("javax.net.ssl")
+          .apply {
+            level = Level.FINEST
+          }
+    }
+  }
+}
+
+class WiresharkExample {
+  val connectionSpec =
+    Builder(ConnectionSpec.RESTRICTED_TLS)
+        .tlsVersions(TLS_1_2)
+//        .cipherSuites(CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256)
```

Comment:
?

Reply:
This was because I believe the choice of non ECDHE cipher might not work.  But I haven't tested.  And I think the widely used ciphers will work.